src/System.Management.Automation/SourceGenerators/PSVersionInfoGenerator/PSVersionInfoGenerator.cs

```diff
@@ -0,0 +1,130 @@
+            const string SourceTemplate = @"// <auto-generated>
+// This file is auto-generated by PSVersionInfoGenerator.
+// </auto-generated>
+
+namespace System.Management.Automation
+{{
+    public static partial class PSVersionInfo
+    {{
+        // Defined in 'PowerShell.Common.props' as 'ProductVersion'
+        // Example:
+        //  - when built from a commit:              ProductVersion = '7.3.0-preview.8 Commits: 29 SHA: 52c6b...'
+        //  - when built from a preview release tag: ProductVersion = '7.3.0-preview.8 SHA: f1ec9...'
+        //  - when built from a stable release tag:  ProductVersion = '7.3.0 SHA: f1ec9...'
+        internal const string ProductVersion = ""{0}"";
+
+        // The git commit id that the build is based off.
+        // Defined in 'PowerShell.Common.props' as 'PowerShellVersion' or 'ReleaseTag',
+        // depending on whether the '-ReleaseTag' is specified when building.
+        // Example:
+        //  - when built from a commit:              GitCommitId = '7.3.0-preview.8-29-g52c6b...'
+        //  - when built from a preview release tag: GitCommitId = '7.3.0-preview.8'
+        //  - when built from a stable release tag:  GitCommitId = '7.3.0'
+        internal const string GitCommitId = ""{1}"";
+
+        // The PowerShell version components.
+        // The version string is defined in 'PowerShell.Common.props' as 'PSCoreBuildVersion',
+        // but we break it into components to save the overhead of parsing at runtime.
+        // Example:
+        //  - '7.3.0-preview.8' for preview release or private build
+        //  - '7.3.0' for stable release
+        private const int Version_Major = {2};
+        private const int Version_Minor = {3};
+        private const int Version_Patch = {4};
+        private const string Version_Label = ""{5}"";
+    }}
+}}";
+
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.ProductVersion", out var productVersion);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.PSCoreBuildVersion", out var mainVersion);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.PowerShellVersion", out var gitDescribe);
+            context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.ReleaseTag", out var releaseTag);
+
+            string gitCommitId = string.IsNullOrEmpty(releaseTag) ? gitDescribe : releaseTag;
+            if (gitCommitId.StartsWith("v"))
+            {
+                gitCommitId = gitCommitId.Substring(1);
+            }
+
+            var result = ParsePSVersion(mainVersion);
+
+            return string.Format(
+                CultureInfo.InvariantCulture,
+                SourceTemplate,
+                productVersion,
+                gitCommitId,
+                result.major,
+                result.minor,
+                result.patch,
+                result.preReleaseLabel);
```

Comment:
Maybe use `string.Create()` and interpolated string. This will make the template more readable.

Reply:
I don't think `string.Create` is available in `netstandard2.0`.
Also, IMO, it's pretty readable already.